File: Sources/ThesisSkipApp/auth0/Auth0Manager.swift
//
//  Auth0Manager.swift
//  MADiOS
//
//  Created by Bram Rampelberg on 03/01/2025.
//  Copyright © 2025 HOGENT. All rights reserved.
//

import Foundation

#if SKIP
    import android.content.Context
    import com.auth0.android.Auth0
    import com.auth0.android.authentication.AuthenticationAPIClient
    import com.auth0.android.authentication.storage.SecureCredentialsManager
    import com.auth0.android.authentication.storage.SharedPreferencesStorage
    import android.content.pm.PackageManager
#else
    import Auth0
#endif

final class Auth0Manager {
    nonisolated(unsafe) static let shared = Auth0Manager()

    let clientId: String
    let domain: String
    let audience: String

    #if SKIP
        let auth0: Auth0

        lazy var androidClient: AuthenticationAPIClient = {
            return AuthenticationAPIClient(auth0)
        }()

        lazy var androidCredentialsManager: SecureCredentialsManager = {
            let context = ProcessInfo.processInfo.androidContext
            return SecureCredentialsManager(
                context,
                auth0,
                SharedPreferencesStorage(context)
            )
        }()
    #else
        lazy var client: Auth0.Authentication = {
            return Auth0.authentication(clientId: clientId, domain: domain)
        }()

        lazy var credentialsManager: CredentialsManager = {
            return CredentialsManager(authentication: client)
        }()
    #endif

    private init() {
        #if SKIP
            let context = ProcessInfo.processInfo.androidContext
            do {
                let appInfo = context.packageManager
                    .getApplicationInfo(
                        context.packageName,
                        PackageManager.GET_META_DATA
                    )
                let metaData = appInfo.metaData

                guard let clientId = metaData?.getString("AUTH0_CLIENT_ID"),
                    let domain = metaData?.getString("AUTH0_DOMAIN"),
                    let audience = metaData?.getString("AUTH0_AUDIENCE")
                else {
                    fatalError(
                        "Auth0 config is missing from AndroidManifest.xml meta-data. Check your skip.env file."
                    )
                }
                self.auth0 = Auth0.getInstance(clientId, domain)
                self.clientId = clientId
                self.domain = domain
                self.audience = audience
            } catch {
                fatalError(
                    "Failed to read AndroidManifest.xml: \(error.localizedDescription)"
                )
            }

        #else
            guard let infoDict = Bundle.main.infoDictionary,
                let clientId = infoDict["AUTH0_CLIENT_ID"] as? String,
                let domain = infoDict["AUTH0_DOMAIN"] as? String,
                let audience = infoDict["AUTH0_AUDIENCE"] as? String
            else {
                fatalError(
                    "Auth0 config is missing or invalid in Info.plist. Check your skip.env file."
                )
            }
            self.clientId = clientId
            self.domain = domain
            self.audience = audience
        #endif
    }
}

File: .build/plugins/outputs/thesis-skip-app/ThesisSkipApp/destination/skipstone/ThesisSkipApp/src/main/kotlin/thesis/skip/app/Auth0Manager.kt
//
//  Auth0Manager.swift
//  MADiOS
//
//  Created by Bram Rampelberg on 03/01/2025.
//  Copyright © 2025 HOGENT. All rights reserved.
//

package thesis.skip.app

import skip.lib.*

import skip.foundation.*

import android.content.Context
import com.auth0.android.Auth0
import com.auth0.android.authentication.AuthenticationAPIClient
import com.auth0.android.authentication.storage.SecureCredentialsManager
import com.auth0.android.authentication.storage.SharedPreferencesStorage
import android.content.pm.PackageManager

internal class Auth0Manager {

    internal val clientId: String
    internal val domain: String
    internal val audience: String

    internal val auth0: Auth0

    internal var androidClient: AuthenticationAPIClient
        get() {
            if (!::androidClientstorage.isInitialized) {
                androidClientstorage = linvoke l@{ -> return@l AuthenticationAPIClient(auth0) }
            }
            return androidClientstorage.sref({ this.androidClient = it })
        }
        set(newValue) {
            androidClientstorage = newValue.sref()
        }
    private lateinit var androidClientstorage: AuthenticationAPIClient

    internal var androidCredentialsManager: SecureCredentialsManager
        get() {
            if (!::androidCredentialsManagerstorage.isInitialized) {
                androidCredentialsManagerstorage = linvoke l@{ ->
                    val context = ProcessInfo.processInfo.androidContext.sref()
                    return@l SecureCredentialsManager(context, auth0, SharedPreferencesStorage(context))
                }
            }
            return androidCredentialsManagerstorage.sref({ this.androidCredentialsManager = it })
        }
        set(newValue) {
            androidCredentialsManagerstorage = newValue.sref()
        }
    private lateinit var androidCredentialsManagerstorage: SecureCredentialsManager

    private constructor() {
        val context = ProcessInfo.processInfo.androidContext.sref()
        try {
            val appInfo = context.packageManager
                .getApplicationInfo(context.packageName, PackageManager.GET_META_DATA)
            val metaData = appInfo.metaData.sref()
            val clientId_0 = metaData?.getString("AUTH0_CLIENT_ID")
            if (clientId_0 == null) {
                fatalError("Auth0 config is missing from AndroidManifest.xml meta-data. Check your skip.env file.")
            }
            val domain_0 = metaData?.getString("AUTH0_DOMAIN")
            if (domain_0 == null) {
                fatalError("Auth0 config is missing from AndroidManifest.xml meta-data. Check your skip.env file.")
            }
            val audience_0 = metaData?.getString("AUTH0_AUDIENCE")
            if (audience_0 == null) {
                fatalError("Auth0 config is missing from AndroidManifest.xml meta-data. Check your skip.env file.")
            }
            this.auth0 = Auth0.getInstance(clientId_0, domain_0)
            this.clientId = clientId_0
            this.domain = domain_0
            this.audience = audience_0
        } catch (error: Throwable) {
            @Suppress("NAME_SHADOWING") val error = error.aserror()
            fatalError("Failed to read AndroidManifest.xml: ${error.localizedDescription}")
        }

    }

    companion object {
        internal val shared = Auth0Manager()
    }
}
